src/core/Akka.Cluster.Tests.MultiNode/QuickRestartSpec.cs

```diff
@@ -0,0 +1,141 @@
+        {
+            var totalNumberOfNodes = Roles.Count + 1;
+            ActorSystem restartingSystem = null; // only used on second
+            for (var i = 1; i <= _rounds; i++)
+            {
+                var round = i; //non-loop variable closure
+                Log.Info("Round-{0}", i);
+                RunOn(() =>
+                {
+                    restartingSystem = restartingSystem == null
+                        ? ActorSystem.Create(Sys.Name, ConfigurationFactory.ParseString($"akka.cluster.roles=[round-{round}]")
+                            .WithFallback(Sys.Settings.Config))
+                        : ActorSystem.Create(Sys.Name, ConfigurationFactory.ParseString($"akka.cluster.roles=[round-{round}]")
+                            .WithFallback(
+                                $"akka.remote.helios.tcp.port={Cluster.Get(restartingSystem).SelfAddress.Port}")
+                            .WithFallback(Sys.Settings.Config));
+                    Log.Info("Restarting node has address {0}", Cluster.Get(restartingSystem).SelfUniqueAddress);
+                    Cluster.Get(restartingSystem).JoinSeedNodes(_seedNodes.Value);
+                    Within(TimeSpan.FromSeconds(20), () =>
+                    {
+                        AwaitAssert(() =>
+                        {
+                            Cluster.Get(restartingSystem).State.Members.Count.ShouldBe(totalNumberOfNodes);
+                            Cluster.Get(restartingSystem).State.Members.All(x => x.Status == MemberStatus.Up).ShouldBeTrue();
+                        });
+                    });
+                }, _config.Second);
+
+                EnterBarrier("joined-"+i);
+                Within(TimeSpan.FromSeconds(20), () =>
+                {
+                    AwaitAssert(() =>
+                    {
+                        Cluster.Get(Sys).State.Members.Count.ShouldBe(totalNumberOfNodes);
+                        Cluster.Get(Sys).State.Members.All(x => x.Status == MemberStatus.Up).ShouldBeTrue();
+                        // use the role to test that it is the new incarnation that joined, sneaky
+                        Cluster.Get(Sys).State.Members.SelectMany(x => x.Roles).ShouldBe(new [] { $"round-{round}" });
+                    });
+                });
+                EnterBarrier("members-up-"+i);
+
+                // gatting occurred after a while
+                if (i > 1)
+                    Thread.Sleep(ThreadLocalRandom.Current.Next(15) * 1000);
+
+                Cluster.Get(Sys).State.Members.Count.ShouldBe(totalNumberOfNodes);
+                Cluster.Get(Sys).State.Members.All(x => x.Status == MemberStatus.Up).ShouldBeTrue();
+                Cluster.Get(Sys).State.Unreachable.Count.ShouldBe(0);
+
+                EnterBarrier("before-terminate-"+i);
+                RunOn(() =>
+                {
+                    restartingSystem.Terminate().Wait(RemainingOrDefault).ShouldBeTrue("Timed out when terminating actorsystem");
```

Comment:
The `ShouldBeTrue` check is innocuous - if the actor system doesn't shut down in time then the test fails anyway. This just helped make it clearer *why* the test failed.

I'll add that missing assertion @maxim-s